### PR TITLE
subscriber: warn if trying to enable a statically disabled level

### PR DESCRIPTION
**Pull request by jyn514**
## Motivation

Fixes https://github.com/tokio-rs/tracing/issues/975.

## Solution

This implements the warning in `tracing-subscriber` only, as mentioned in https://github.com/tokio-rs/tracing/issues/975#issuecomment-692903469. It warns whenever directives are parsed, including programmatically and through environment variables. It does not include the suggested new API which returns the filters that wouldn't be parsed.

- List filters that would be ignored
- Mention 'static max level'
- Describe how to enable the logging

Example output:

```
$ RUST_LOG=off,debug,silenced[x]=trace cargo run -q
warning: some trace filter directives would enable traces that are disabled statically
 | `debug` would enable the DEBUG level for all targets
 | `silenced[x]=trace` would enable the TRACE level for the `silenced` target
 = note: the static max level is info
 = note: to enable DEBUG logging, remove the `max_level_info` feature
```
image

### CODE REVIEW

**Comment by jyn514** [comment]
I'm not sure how to add tests for this ... the existing tests in `tracing-subscriber/tests` use mocks and don't touch stdout. Should I add some example binary crates and run those?

**Comment by jyn514** [comment]
Originally this added a `Copy` impl for `Level` and `LevelFilter`, but that ran into a ton of warnings so I'm going to split it into a separate PR.

File: tracing-subscriber/Cargo.toml [hunk 1]
```diff
@@ -24,7 +24,7 @@ keywords = ["logging", "tracing", "metrics", "subscriber"]
 [features]
 
 default = ["env-filter", "smallvec", "fmt", "ansi", "chrono", "tracing-log", "json"]
-env-filter = ["matchers", "regex", "lazy_static"]
+env-filter = ["matchers", "regex", "lazy_static", "tracing"]
```
Comment (davidbarsky) [inline tracing-subscriber/Cargo.toml:27]: Not sure `tracing` needs to be another feature flag on this. I don't see this flag being used elsewhere _and_ I think that a tracing dependency is a forgone conclusion for `tracing-subscriber`.

File: tracing-subscriber/src/filter/env/directive.rs [hunk 2]
```diff
@@ -9,10 +9,10 @@ use tracing_core::{span, Level, Metadata};
 // TODO(eliza): add a builder for programmatically constructing directives?
 #[derive(Debug, Eq, PartialEq)]
 pub struct Directive {
-    target: Option<String>,
```
Comment (davidbarsky) [inline tracing-subscriber/src/filter/env/directive.rs:12]: Why the change in visibility?

File: tracing-subscriber/src/filter/env/mod.rs [hunk 3]
```diff
@@ -242,6 +242,55 @@ impl EnvFilter {
...
+
+        let directives: Vec<_> = directives.into_iter().collect();
+
+        let disabled: Vec<_> = directives
+            .iter()
+            .filter(|directive| directive.level > STATIC_MAX_LEVEL)
+            .collect();
+
+        if !disabled.is_empty() {
+            eprintln!("warning: some trace filter directives would enable traces that are disabled statically");
```
Comment (davidbarsky) [inline tracing-subscriber/src/filter/env/mod.rs:256]: Nit: can you add a comment explaining why `stderr` was chosen? E.g., I'm assuming that the EnvFilter can't access the underlying `MakeWriter`, let alone assume one is present.

File: tracing-subscriber/src/filter/env/directive.rs (hunk 2 above)
Comment (jyn514) [reply tracing-subscriber/src/filter/env/directive.rs:12]: Otherwise it won't compile:

```
error[E0616]: field `target` of struct `filter::env::directive::Directive` is private
   --> tracing-subscriber/src/filter/env/mod.rs:258:63
    |
258 |                 let target = if let Some(target) = &directive.target {
    |                                                               ^^^^^^ private field
```

File: tracing-subscriber/Cargo.toml (hunk 1 above)
Comment (jyn514) [reply tracing-subscriber/Cargo.toml:27]: This won't compile without adding a `tracing` dependency:

```
error[E0433]: failed to resolve: use of undeclared type or module `tracing`
   --> tracing-subscriber/src/filter/env/mod.rs:245:13
    |
245 |         use tracing::level_filters::STATIC_MAX_LEVEL;
    |             ^^^^^^^ use of undeclared type or module `tracing`
error[E0433]: failed to resolve: use of undeclared type or module `tracing`
   --> tracing-subscriber/src/filter/env/mod.rs:245:13
    |
245 |         use tracing::level_filters::STATIC_MAX_LEVEL;
    |             ^^^^^^^ use of undeclared type or module `tracing`
```

File: tracing-subscriber/src/filter/env/directive.rs (hunk 2 above)
Comment (davidbarsky) [reply tracing-subscriber/src/filter/env/directive.rs:12]: Ah, fair. In that case, making this `pub(crate)` is reasonable to me.

From a (very!) pedantic style/code formatting point of view, it'd be nice to have the pub(crate) fields grouped together at the end of the struct.

File: tracing-subscriber/Cargo.toml (hunk 1 above)
Comment (davidbarsky) [reply tracing-subscriber/Cargo.toml:27]: I'd like to hear what @hawkw says, but I'm not sure `tracing` merits being an optional dependency.

File: tracing-subscriber/src/filter/env/mod.rs [hunk 4]
```diff
@@ -258,6 +258,8 @@ impl EnvFilter {
             .collect();
 
         if !disabled.is_empty() {
+            // NOTE: this can't use `MakeWriter` because it's not yet available;
+            // there is not yet a subscriber.
```
Comment (davidbarsky) [inline tracing-subscriber/src/filter/env/mod.rs:262]:
```suggestion
            // NOTE: We can't use a configured `MakeWriter` because the EnvFilter
            // has no knowledge of any underlying subscriber or collector, which 
            // may or may not use a `MakeWriter`.
```
(Terminology borrowed from #1015)

File: tracing-subscriber/Cargo.toml (hunk 1 above)
Comment (jyn514) [reply tracing-subscriber/Cargo.toml:27]: Oh I see, you want to make it always a dependency instead of optional. Sure, I can do that.

**Comment by jyn514** [comment]
Rebased over master to get the test fix from #1016.

File: tracing-subscriber/Cargo.toml (hunk 1 above)
Comment (hawkw) [reply tracing-subscriber/Cargo.toml]: I think the `tracing` dependency should be optional and only enabled by the "env-filter" feature. Otherwise, there's no reason to depend on `tracing` --- the rest of the crate only uses `tracing-core`.

I think @davidbarsky misinterpreted this change as adding a new _feature flag_ "tracing" that users could independently enable. That's not what this change is doing --- it simply makes the "env-filter" feature enable a dependency on `tracing` that isn't otherwise necessary.

IMO the original code was correct.

File: tracing-subscriber/src/filter/env/mod.rs (hunk 3 above)
Comment (hawkw) [reply tracing-subscriber/src/filter/env/mod.rs:264]: fwiw, this function already printed to stderr when directives don't parse correctly (borrowed from `env-filter`)

File: tracing-subscriber/src/filter/env/mod.rs [hunk 5]
```diff
@@ -242,6 +242,55 @@ impl EnvFilter {
...
+                    Some(Level::ERROR) => ("max_level_error", Level::WARN),
+                    None => return ("max_level_off", String::new()),
+                };
+                (feature, format!("{} ", filter))
+            };
+            let (feature, earlier_level) = help_msg();
+            eprintln!(
+                "note: to enable {}logging, remove the `{}` feature",
+                earlier_level, feature
+            );
```
[the remark this answers is not in the document]
Comment (jyn514) [reply tracing-subscriber/src/filter/env/mod.rs:299]: What do you think?

colors

[the remark this answers is not in the document]
Comment (jyn514) [reply tracing-subscriber/src/filter/env/mod.rs:299]: For context, this is what rustc's diagnostics look like:

rustc

[the remark this answers is not in the document]
Comment (jyn514) [reply tracing-subscriber/src/filter/env/mod.rs]: > i was trying to imitate the formatting used by rustc's error messages, since I know this feature was originally requested for rustc.

I should probably work on colorizing bootstraps errors ... right now they're pretty inconsistent.

```
(-bash@build-server) ~/rustc $ rg println src/bootstrap/
src/bootstrap/config.rs
565:                    println!("failed to parse TOML configuration '{}': {}", file.display(), err);

src/bootstrap/setup.rs
81:                println!("error: unrecognized option '{}'", input.trim());
82:                println!("note: press Ctrl+C to exit");
```

File: tracing-subscriber/Cargo.toml (hunk 1 above)
Comment (jyn514) [reply tracing-subscriber/Cargo.toml:27]: Ok, I reverted it to the original code.

File: tracing-subscriber/src/filter/env/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (hawkw) [reply tracing-subscriber/src/filter/env/mod.rs]: > What do you think?
> 
> colors

this is beautiful <3

[the remark this answers is not in the document]
Comment (hawkw) [reply tracing-subscriber/src/filter/env/mod.rs]: i _might_ format "the static max level is XXX" as a `note:` but take it or leave it.

[the remark this answers is not in the document]
Comment (jyn514) [reply tracing-subscriber/src/filter/env/mod.rs]: note

[the remark this answers is not in the document]
Comment (davidbarsky) [reply tracing-subscriber/src/filter/env/mod.rs]: this looks rad!

**Comment by jyn514** [comment]
Is this waiting on anything? I'd love to have it in rustc :)

**Comment by jyn514** [comment]
I can squash the commit history if you like, not sure what the policy is for that.

**Review by hawkw — APPROVED** [review]
looks good to me!

**Comment by hawkw** [comment]
> Is this waiting on anything? I'd love to have it in rustc :)

Nope, everything looks fine, I just had a bunch of other things to take care of. Happy to merge it now!

Note that the `master` branch currently has breaking changes for the upcoming`tracing` 0.2.x. If you want to see this in `rustc` tomorrow, we'll need to backport this to the [`v0.1.x` branch](https://github.com/tokio-rs/tracing/tree/v0.1.x) as well. I'd recommend doing that by cherry-picking the merged commit on master.

> I can squash the commit history if you like, not sure what the policy is for that.

We'll squash the PR when merging it. See https://github.com/tokio-rs/tracing/blob/master/CONTRIBUTING.md#commit-squashing :)
